### ISSUE TITLE
PIP installation fails (Win10)

### ISSUE THREAD

**Issue by lcnittl** [codypiersall/pynng#8]
Installation of `pynng` via `pip` fails for my Win10 machine (Version	10.0.17134 Build 17134). Before being able to install at all, I had to run `pip install cmake` (otherwise, `cmake` would be an unrecognized command).

Is the problem a wrong/incomplete installation of `Visual Studio 14 2015 Win64`, or is there some error in the `setup.py`?

The output of the command prompt (`pip` is run as administrator):
```
C:\>pip install pynng
Collecting pynng
  Using cached https://files.pythonhosted.org/packages/43/5f/2b31533f9d44fe568a49bb309ae36f3c8e32a62008826939191885b46b94/pynng-0.1.2.tar.gz
Requirement already satisfied: cffi in c:\program files\python37\lib\site-packages (from pynng) (1.11.5)
Requirement already satisfied: pycparser in c:\program files\python37\lib\site-packages (from cffi->pynng) (2.19)
Installing collected packages: pynng
  Running setup.py install for pynng ... error
    Complete output from command "c:\program files\python37\python.exe" -u -c "import setuptools, tokenize;__file__='C:\\Users\\lcnittl\\AppData\\Local\\Temp\\pip-install-0b5n25ym\\pynng\\setup.py';f=getattr(tokenize, 'open', open)(__file__);code=f.read().replace('\r\n', '\n');f.close();exec(compile(code, __file__, 'exec'))" install --record C:\Users\lcnittl\AppData\Local\Temp\pip-record-o72gsi5f\install-record.txt --single-version-externally-managed --compile:
    Warning: 'classifiers' should be a list, got type 'tuple'
    running install
    running build
    running build_py

    C:\Users\lcnittl\AppData\Local\Temp\pip-install-0b5n25ym\pynng>if {"Visual Studio 14 2015 Win64"} == {} (
    echo.ERROR: You must provide the correct CMake compiler generator.
     goto :end
    )

    C:\Users\lcnittl\AppData\Local\Temp\pip-install-0b5n25ym\pynng>pushd .

    C:\Users\lcnittl\AppData\Local\Temp\pip-install-0b5n25ym\pynng>rmdir /s /q nng\build
    The system cannot find the file specified.

    C:\Users\lcnittl\AppData\Local\Temp\pip-install-0b5n25ym\pynng>if not exist nng (
    git clone https://github.com/nanomsg/nng nng
     pushd nng
     git checkout 6c334f3
     popd
    )

    C:\Users\lcnittl\AppData\Local\Temp\pip-install-0b5n25ym\pynng>mkdir nng\build

    C:\Users\lcnittl\AppData\Local\Temp\pip-install-0b5n25ym\pynng>cd nng\build   && cmake -G "Visual Studio 14 2015 Win64" ..   && cmake --build . --config Release
    -- Selecting Windows SDK version  to target Windows 10.0.17134.
    -- The C compiler identification is MSVC 19.0.24210.0
    -- Check for working C compiler: C:/Program Files (x86)/Microsoft Visual Studio 14.0/VC/bin/x86_amd64/cl.exe
    -- Check for working C compiler: C:/Program Files (x86)/Microsoft Visual Studio 14.0/VC/bin/x86_amd64/cl.exe -- works
    -- Detecting C compiler ABI info
    -- Detecting C compiler ABI info - done
    -- Detecting C compile features
    -- Detecting C compile features - done
    -- Could NOT find UnixCommands (missing: BASH CP MV RM)
    -- Check if the system is big endian
    -- Searching 16 bit integer
    -- Looking for sys/types.h
    -- Looking for sys/types.h - found
    -- Looking for stdint.h
    -- Looking for stdint.h - found
    -- Looking for stddef.h
    -- Looking for stddef.h - found
    -- Check size of unsigned short
    -- Check size of unsigned short - done
    -- Using unsigned short
    -- Check if the system is big endian - little endian
    -- Looking for pthread.h
    -- Looking for pthread.h - not found
    -- Found Threads: TRUE
    -- Looking for InitializeConditionVariable
    -- Looking for InitializeConditionVariable - found
    -- Looking for snprintf
    -- Looking for snprintf - found
    -- Looking for strlcat
    -- Looking for strlcat - not found
    -- Looking for strlcpy
    -- Looking for strlcpy - not found
    -- Looking for strnlen
    -- Looking for strnlen - found
    -- Looking for strcasecmp
    -- Looking for strcasecmp - not found
    -- Looking for strncasecmp
    -- Looking for strncasecmp - not found
    -- Building static libs
    -- Test tls disabled (unconfigured)
    -- Test wss disabled (unconfigured)
    -- Test wssfile disabled (unconfigured)
    -- Test zt disabled (unconfigured)
    -- The CXX compiler identification is MSVC 19.0.24210.0
    -- Check for working CXX compiler: C:/Program Files (x86)/Microsoft Visual Studio 14.0/VC/bin/x86_amd64/cl.exe
    -- Check for working CXX compiler: C:/Program Files (x86)/Microsoft Visual Studio 14.0/VC/bin/x86_amd64/cl.exe -- works
    -- Detecting CXX compiler ABI info
    -- Detecting CXX compiler ABI info - done
    -- Detecting CXX compile features
    -- Detecting CXX compile features - done
    CMake Error at CMakeLists.txt:497 (add_subdirectory):
      add_subdirectory given source "docs/man" which is not an existing
      directory.


    -- Configuring incomplete, errors occurred!
    See also "C:/Users/lcnittl/AppData/Local/Temp/pip-install-0b5n25ym/pynng/nng/build/CMakeFiles/CMakeOutput.log".
    See also "C:/Users/lcnittl/AppData/Local/Temp/pip-install-0b5n25ym/pynng/nng/build/CMakeFiles/CMakeError.log".

    C:\Users\lcnittl\AppData\Local\Temp\pip-install-0b5n25ym\pynng\nng\build>popd
    creating build
    creating build\lib.win-amd64-3.7
    creating build\lib.win-amd64-3.7\pynng
    copying pynng\exceptions.py -> build\lib.win-amd64-3.7\pynng
    copying pynng\nng.py -> build\lib.win-amd64-3.7\pynng
    copying pynng\__init__.py -> build\lib.win-amd64-3.7\pynng
    creating build\lib.win-amd64-3.7\test
    copying test\test_api.py -> build\lib.win-amd64-3.7\test
    copying test\test_options.py -> build\lib.win-amd64-3.7\test
    copying test\test_protocols.py -> build\lib.win-amd64-3.7\test
    copying test\__init__.py -> build\lib.win-amd64-3.7\test
    running build_ext
    generating cffi module 'build\\temp.win-amd64-3.7\\Release\\pynng._nng.c'
    creating build\temp.win-amd64-3.7
    creating build\temp.win-amd64-3.7\Release

    C:\Users\lcnittl\AppData\Local\Temp\pip-install-0b5n25ym\pynng>if {"Visual Studio 14 2015 Win64"} == {} (
    echo.ERROR: You must provide the correct CMake compiler generator.
     goto :end
    )

    C:\Users\lcnittl\AppData\Local\Temp\pip-install-0b5n25ym\pynng>pushd .

    C:\Users\lcnittl\AppData\Local\Temp\pip-install-0b5n25ym\pynng>rmdir /s /q nng\build

    C:\Users\lcnittl\AppData\Local\Temp\pip-install-0b5n25ym\pynng>if not exist nng (
    git clone https://github.com/nanomsg/nng nng
     pushd nng
     git checkout 6c334f3
     popd
    )

    C:\Users\lcnittl\AppData\Local\Temp\pip-install-0b5n25ym\pynng>mkdir nng\build

    C:\Users\lcnittl\AppData\Local\Temp\pip-install-0b5n25ym\pynng>cd nng\build   && cmake -G "Visual Studio 14 2015 Win64" ..   && cmake --build . --config Release
    -- Selecting Windows SDK version  to target Windows 10.0.17134.
    -- The C compiler identification is MSVC 19.0.24210.0
    -- Check for working C compiler: C:/Program Files (x86)/Microsoft Visual Studio 14.0/VC/bin/x86_amd64/cl.exe
    -- Check for working C compiler: C:/Program Files (x86)/Microsoft Visual Studio 14.0/VC/bin/x86_amd64/cl.exe -- works
    -- Detecting C compiler ABI info
    -- Detecting C compiler ABI info - done
    -- Detecting C compile features
    -- Detecting C compile features - done
    -- Could NOT find UnixCommands (missing: BASH CP MV RM)
    -- Check if the system is big endian
    -- Searching 16 bit integer
    -- Looking for sys/types.h
    -- Looking for sys/types.h - found
    -- Looking for stdint.h
    -- Looking for stdint.h - found
    -- Looking for stddef.h
    -- Looking for stddef.h - found
    -- Check size of unsigned short
    -- Check size of unsigned short - done
    -- Using unsigned short
    -- Check if the system is big endian - little endian
    -- Looking for pthread.h
    -- Looking for pthread.h - not found
    -- Found Threads: TRUE
    -- Looking for InitializeConditionVariable
    -- Looking for InitializeConditionVariable - found
    -- Looking for snprintf
    -- Looking for snprintf - found
    -- Looking for strlcat
    -- Looking for strlcat - not found
    -- Looking for strlcpy
    -- Looking for strlcpy - not found
    -- Looking for strnlen
    -- Looking for strnlen - found
    -- Looking for strcasecmp
    -- Looking for strcasecmp - not found
    -- Looking for strncasecmp
    -- Looking for strncasecmp - not found
    -- Building static libs
    -- Test tls disabled (unconfigured)
    -- Test wss disabled (unconfigured)
    -- Test wssfile disabled (unconfigured)
    -- Test zt disabled (unconfigured)
    -- The CXX compiler identification is MSVC 19.0.24210.0
    -- Check for working CXX compiler: C:/Program Files (x86)/Microsoft Visual Studio 14.0/VC/bin/x86_amd64/cl.exe
    -- Check for working CXX compiler: C:/Program Files (x86)/Microsoft Visual Studio 14.0/VC/bin/x86_amd64/cl.exe -- works
    -- Detecting CXX compiler ABI info
    -- Detecting CXX compiler ABI info - done
    -- Detecting CXX compile features
    -- Detecting CXX compile features - done
    CMake Error at CMakeLists.txt:497 (add_subdirectory):
      add_subdirectory given source "docs/man" which is not an existing
      directory.


    -- Configuring incomplete, errors occurred!
    See also "C:/Users/lcnittl/AppData/Local/Temp/pip-install-0b5n25ym/pynng/nng/build/CMakeFiles/CMakeOutput.log".
    See also "C:/Users/lcnittl/AppData/Local/Temp/pip-install-0b5n25ym/pynng/nng/build/CMakeFiles/CMakeError.log".

    C:\Users\lcnittl\AppData\Local\Temp\pip-install-0b5n25ym\pynng\nng\build>popd
    building 'pynng._nng' extension
    creating build\temp.win-amd64-3.7\Release\build
    creating build\temp.win-amd64-3.7\Release\build\temp.win-amd64-3.7
    creating build\temp.win-amd64-3.7\Release\build\temp.win-amd64-3.7\Release
    C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\BIN\x86_amd64\cl.exe /c /nologo /Ox /W3 /GL /DNDEBUG /MD -Inng/src "-Ic:\program files\python37\include" "-Ic:\program files\python37\include" "-IC:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\INCLUDE" "-IC:\Program Files (x86)\Windows Kits\10\include\10.0.10240.0\ucrt" "-IC:\Program Files (x86)\Windows Kits\8.1\include\shared" "-IC:\Program Files (x86)\Windows Kits\8.1\include\um" "-IC:\Program Files (x86)\Windows Kits\8.1\include\winrt" /Tcbuild\temp.win-amd64-3.7\Release\pynng._nng.c /Fobuild\temp.win-amd64-3.7\Release\build\temp.win-amd64-3.7\Release\pynng._nng.obj
    pynng._nng.c
    C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\BIN\x86_amd64\link.exe /nologo /INCREMENTAL:NO /LTCG /DLL /MANIFEST:EMBED,ID=2 /MANIFESTUAC:NO "/LIBPATH:c:\program files\python37\libs" "/LIBPATH:c:\program files\python37\PCbuild\amd64" "/LIBPATH:C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\LIB\amd64" "/LIBPATH:C:\Program Files (x86)\Windows Kits\10\lib\10.0.10240.0\ucrt\x64" "/LIBPATH:C:\Program Files (x86)\Windows Kits\8.1\lib\winv6.3\um\x64" Ws2_32.lib Advapi32.lib /EXPORT:PyInit__nng build\temp.win-amd64-3.7\Release\build\temp.win-amd64-3.7\Release\pynng._nng.obj ./nng/build/Release/nng.lib /OUT:build\lib.win-amd64-3.7\pynng\_nng.cp37-win_amd64.pyd /IMPLIB:build\temp.win-amd64-3.7\Release\build\temp.win-amd64-3.7\Release\_nng.cp37-win_amd64.lib
    LINK : fatal error LNK1181: cannot open input file './nng/build/Release/nng.lib'
    error: command 'C:\\Program Files (x86)\\Microsoft Visual Studio 14.0\\VC\\BIN\\x86_amd64\\link.exe' failed with exit status 1181

    ----------------------------------------
Command ""c:\program files\python37\python.exe" -u -c "import setuptools, tokenize;__file__='C:\\Users\\lcnittl\\AppData\\Local\\Temp\\pip-install-0b5n25ym\\pynng\\setup.py';f=getattr(tokenize, 'open', open)(__file__);code=f.read().replace('\r\n', '\n');f.close();exec(compile(code, __file__, 'exec'))" install --record C:\Users\lcnittl\AppData\Local\Temp\pip-record-o72gsi5f\install-record.txt --single-version-externally-managed --compile" failed with error code 1 in C:\Users\lcnittl\AppData\Local\Temp\pip-install-0b5n25ym\pynng\
```

Thanks in advance for any help!

**Comment by lcnittl** [comment]
Installation seemed to work using the "manual" install via command line.

```
git clone https://github.com/codypiersall/pynng
cd pynng
pip install .
python setup.py build
python setup.py build_ext --inplace
pytest
```

**Comment by codypiersall** [comment]
Thanks for filing the issue.  There were a couple problems here, one I knew about and one I didn't:

1. First the one I did not know about: last time I was uploading the wheels to PyPI, I neglected to upload the 64-bit Windows Python 3.7 build, so pip couldn't find the latest version on PyPI.  I have fixed this now, so no one should run into this.
2. Since pip couldn't find the latest binary build, it downloaded the source distribution, which unfortunately was broken.  It was fixed in db299bf2b6e137881a28c37869ce06fbba9a8362, but I haven't uploaded a new release yet; I'm planning on bumping it when #4  is completed.

I should also note that cmake is required when building form source in the README.

P. S. I had _no idea_ that cmake could be installed with pip.  Looks like I could have avoided building my own Docker image for Travis CI... oh well!
